File: agent-platform-api/app/schemas/builder.py
from pydantic import BaseModel, Field


class BuilderRequest(BaseModel):
    prompt: str = Field(..., min_length=10)
    conversation_history: list[dict] = Field(default_factory=list)
    current_definition: str | None = None
    context: dict | None = None

    model_config = {"from_attributes": True}


class BuilderResponse(BaseModel):
    definition_md: str
    explanation: str
    tools_used: list[str]
    warnings: list[str]
    suggestions: list[str]

    model_config = {"from_attributes": True}


class ValidateRequest(BaseModel):
    definition_md: str = Field(..., min_length=10)

    model_config = {"from_attributes": True}


class ValidationResult(BaseModel):
    is_valid: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    suggestions: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class SimulationRequest(BaseModel):
    agent_id: str | None = None
    definition_md: str | None = None
    mock_inputs: dict = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class SimulationResult(BaseModel):
    status: str
    timeline: list[dict] = Field(default_factory=list)
    output: dict | None = None
    tokens_used: int = 0
    warnings: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}

File: agent-platform-ui/src/api/builder.ts
import apiClient from './client';

export interface BuilderGenerateRequest {
  prompt: string;
  conversation_history: Array<{ role: string; content: string }>;
  current_definition: string | null;
  context?: Record<string, unknown>;
}

export interface BuilderGenerateResponse {
  definition_md: string;
  explanation: string;
  tools_used: string[];
  warnings: string[];
  suggestions: string[];
}

export interface ValidateRequest {
  definition_md: string;
}

export interface ValidationResult {
  is_valid: boolean;
  errors: string[];
  warnings: string[];
  suggestions: string[];
}

export interface SimulationRequest {
  agent_id?: string;
  definition_md?: string;
  mock_inputs: Record<string, unknown>;
}

export interface SimulationTimelineEntry {
  step: number;
  action: string;
  status: string;
  message?: string;
  tool_name?: string;
  duration_ms: number;
  timestamp: string;
  tool_call?: {
    tool_name: string;
    input_summary: string;
    input: Record<string, unknown>;
    output: Record<string, unknown>;
    status: string;
    duration_ms: number;
  };
}

export interface SimulationResult {
  status: string;
  timeline: SimulationTimelineEntry[];
  output: Record<string, unknown> | null;
  tokens_used: number;
  warnings: string[];
}

export async function generateAgent(
  data: BuilderGenerateRequest,
): Promise<BuilderGenerateResponse> {
  const response = await apiClient.post<BuilderGenerateResponse>(
    '/api/builder/generate',
    data,
  );
  return response.data;
}

export async function validateDefinition(
  definitionMd: string,
): Promise<ValidationResult> {
  const response = await apiClient.post<ValidationResult>(
    '/api/builder/validate',
    { definition_md: definitionMd },
  );
  return response.data;
}

export async function simulateAgent(
  data: SimulationRequest,
): Promise<SimulationResult> {
  const response = await apiClient.post<SimulationResult>(
    '/api/builder/simulate',
    data,
  );
  return response.data;
}
